File: blue/src/package_neon_multi_node_blue/storage.py
"""Deployment-owned S3 application buckets and scoped credentials."""
import re
import json
from pathlib import Path
from blue import tofu
from blue.cli import stage_dir
from blue.runtime import runtime
from blue.scaffold import scaffold, PRESERVE_JINJA_DELIMITERS

tool = "neon-multi-node-storage"


def managed(opts):
    return opts.get("neon-storage-managed") is True


def directory(opts):
    return stage_dir(opts, tool, default_profile="neon-multi-node")


def aws_env(opts):
    mapping = {"aws-access-key-id": "AWS_ACCESS_KEY_ID", "aws-secret-access-key": "AWS_SECRET_ACCESS_KEY", "aws-session-token": "AWS_SESSION_TOKEN"}
    return {variable: str(opts[key]) for key, variable in mapping.items() if opts.get(key)}


def specs(opts):
    return [{"template": {"name": "tools/storage/main.tf", "content": (Path(__file__).parent / "resources/tools/storage/main.tf").read_text()}, "target": directory(opts) + "/main.tf", "data": opts, "opts": PRESERVE_JINJA_DELIMITERS}]


async def ownership_preflight(opts):
    config = {"cwd": directory(opts), "env": aws_env(opts)}
    result = await runtime.exec(["tofu", "init", "-input=false", "-no-color"], **config)
    if result.exit:
        raise RuntimeError("managed storage state operation failed")
    result = await runtime.exec(["tofu", "state", "list"], **config)
    if result.exit and not (result.exit == 1 and "No state file was found!" in (result.err or "")):
        raise RuntimeError("managed storage state operation failed")
    resources = []
    if not result.exit and result.out.strip():
        shown = await runtime.exec(["tofu", "show", "-json"], **config)
        if shown.exit:
            raise RuntimeError("managed storage state operation failed")
        resources = json.loads(shown.out).get("values", {}).get("root_module", {}).get("resources", [])
    for role, key in [("neon", "neon-r2-bucket")]:
        if any(resource.get("address") == f'aws_s3_bucket.application["{role}"]' and resource.get("values", {}).get("bucket") == opts[key] for resource in resources):
            continue
        probe = await runtime.exec(["aws", "s3api", "head-bucket", "--bucket", opts[key], "--region", opts["neon-r2-region"]], **config)
        if not (probe.exit > 0 and re.search(r"\(404\)|Not Found|NoSuchBucket", probe.err or "")):
            raise RuntimeError("managed storage refuses to adopt an existing or inaccessible bucket")


def writer_proof(opts):
    from .topology import topology
    from colors_compute.contract import expand
    return {n['node_id'] for n in expand(topology(opts))} == opts.get('neon-multi-node/writers-stopped')

async def storage_step(opts):
    if not managed(opts):
        return {**opts, "blue/exit": 0}
    try:
        if opts.get('blue/event')=='delete' and not opts.get('blue/dry-run') and not writer_proof(opts):
            raise RuntimeError('current-run writer stop proof required before storage deletion')
        documents = specs(opts)
        if opts.get("blue/event") == "create":
            scaffold(opts, documents)
            await ownership_preflight(opts)
        return await tofu.tofu_with_spec(opts, documents, dir=directory(opts), env=aws_env(opts), output_key="neon-multi-node/storage-credentials")
    except Exception:
        return {**opts, "blue/exit": 1, "blue/err": "managed S3 storage failed; inspect bucket ownership, state access, and AWS permissions"}


def credential_env(opts):
    credentials = opts.get("neon-multi-node/storage-credentials", {}).get("credentials", {})
    environment = {}
    for role, prefix in [("neon", "NEON_R2"), ("neon", "NEON_S3")]:
        values = credentials.get(role, {})
        access, secret = values.get("access_key_id"), values.get("secret_access_key")
        if not access or not secret or not access.strip() or not secret.strip():
            raise RuntimeError("managed storage credentials unavailable")
        environment[f"COLORS_PAR_{prefix}_ACCESS_KEY_ID"] = access
        environment[f"COLORS_PAR_{prefix}_SECRET_ACCESS_KEY"] = secret
    return environment


async def read_credentials(opts):
    if not managed(opts):
        return opts
    try:
        tofu.conventional_backend_advice(dir=directory, key=lambda o: f"{o['profile']}/{tool}.tfstate")(opts)
        scaffold({**opts, "blue/event": "build"}, specs(opts))
        initialized = await runtime.exec(["tofu", "init", "-input=false", "-no-color"], cwd=directory(opts), env=aws_env(opts))
        if initialized.exit:
            raise RuntimeError("initialization failed")
        result = {**opts, "neon-multi-node/storage-credentials": await tofu.outputs(directory(opts), aws_env(opts))}
        credential_env(result)
        return result
    except Exception:
        raise RuntimeError("managed storage credentials unavailable; converge storage before rehearsal") from None

File: red/src/storage.ts
// Deployment-owned S3 application buckets and scoped credentials.
import {stageDir} from "red/cli";
import {runtime} from "red/runtime";
import {scaffold, PRESERVE_JINJA_DELIMITERS, type Spec} from "red/scaffold";
import * as tofu from "red/tofu";
import type {Opts} from "red/workflow";
import {expand} from "colors-compute-red";
import {topology} from "./topology.ts";
import mainTf from "../resources/tools/storage/main.tf" with {type: "text"};

export const tool = "neon-multi-node-storage";
export const managed = (opts: Opts): boolean => opts["neon-storage-managed"] === true;
export const directory = (opts: Opts): string => stageDir(opts, tool, {defaultProfile: "neon-multi-node"});
export function awsEnv(opts: Opts): Record<string, string> {
  return Object.fromEntries(Object.entries({"aws-access-key-id":"AWS_ACCESS_KEY_ID", "aws-secret-access-key":"AWS_SECRET_ACCESS_KEY", "aws-session-token":"AWS_SESSION_TOKEN"}).filter(([key]) => opts[key]).map(([key, variable]) => [variable, String(opts[key])]));
}
export function specs(opts: Opts): Spec[] {
  return [{template:{name:"tools/storage/main.tf", content:mainTf}, target:directory(opts)+"/main.tf", data:opts, opts:PRESERVE_JINJA_DELIMITERS}];
}
export async function ownershipPreflight(opts: Opts): Promise<void> {
  const config = {cwd:directory(opts), env:awsEnv(opts)};
  const init = await runtime.exec(["tofu", "init", "-input=false", "-no-color"], config);
  if (init.exit !== 0) throw new Error("managed storage state operation failed");
  const state = await runtime.exec(["tofu", "state", "list"], config);
  if (state.exit !== 0 && !(state.exit === 1 && /No state file was found!/.test(state.err))) throw new Error("managed storage state operation failed");
  let resources: Array<{address:string, values?:{bucket?:string}}> = [];
  if (state.exit === 0 && state.out.trim()) {
    const shown = await runtime.exec(["tofu", "show", "-json"], config);
    if (shown.exit !== 0) throw new Error("managed storage state operation failed");
    resources = JSON.parse(shown.out).values?.root_module?.resources ?? [];
  }
  for (const [role,key] of [["neon","neon-r2-bucket"]]) {
    if (resources.some(resource => resource.address === `aws_s3_bucket.application["${role}"]` && resource.values?.bucket === opts[key!])) continue;
    const probe = await runtime.exec(["aws","s3api","head-bucket","--bucket",String(opts[key!]),"--region",String(opts["neon-r2-region"])],config);
    if (!(probe.exit > 0 && /\(404\)|Not Found|NoSuchBucket/.test(probe.err))) throw new Error("managed storage refuses to adopt an existing or inaccessible bucket");
  }
}
export async function storageStep(opts: Opts): Promise<Opts> {
  if (!managed(opts)) return {...opts,"red/exit":0};
  try {
    const documents = specs(opts);
    if(opts["red/event"]==="delete"&&!opts["red/dry-run"]&&!writerProof(opts))throw Error("current-run writer stop proof required before storage deletion");
    if (opts["red/event"] === "create") { scaffold(opts,documents); await ownershipPreflight(opts); }
    return await tofu.tofuWithSpec(opts,documents,{dir:directory(opts),env:awsEnv(opts),outputKey:"neon-multi-node/storage-credentials"});
  } catch { return {...opts,"red/exit":1,"red/err":"managed S3 storage failed; inspect bucket ownership, state access, and AWS permissions"}; }
}
export function credentialEnv(opts: Opts): Record<string,string> {
  const credentials = opts["neon-multi-node/storage-credentials"]?.credentials ?? {};
  const environment: Record<string,string> = {};
  for (const [role,prefix] of [["neon","NEON_R2"],["neon","NEON_S3"]]) {
    const {access_key_id,secret_access_key} = credentials[role!] ?? {};
    if (!String(access_key_id??"").trim() || !String(secret_access_key??"").trim()) throw new Error("managed storage credentials unavailable");
    environment[`COLORS_PAR_${prefix}_ACCESS_KEY_ID`] = access_key_id;
    environment[`COLORS_PAR_${prefix}_SECRET_ACCESS_KEY`] = secret_access_key;
  }
  return environment;
}

export async function readCredentials(opts: Opts): Promise<Opts> {
  if (!managed(opts)) return opts;
  try {
    tofu.conventionalBackendAdvice({dir:directory, key:o => `${o.profile}/${tool}.tfstate`})(opts);
    scaffold({...opts,"red/event":"build"}, specs(opts));
    const initialized = await runtime.exec(["tofu","init","-input=false","-no-color"], {cwd:directory(opts),env:awsEnv(opts)});
    if (initialized.exit !== 0) throw new Error("initialization failed");
    const result = {...opts,"neon-multi-node/storage-credentials":await tofu.outputs(directory(opts),awsEnv(opts))};
    credentialEnv(result);
    return result;
  } catch { throw new Error("managed storage credentials unavailable; converge storage before rehearsal"); }
}

export function writerProof(opts:Opts):boolean {const expected=expand(topology(opts)).map(n=>n.node_id).sort();const actual=opts["neon-multi-node/writers-stopped"];return actual instanceof Set&&JSON.stringify([...actual].sort())===JSON.stringify(expected);}
